File: 149. Max Points on a Line/C++/Solution.cpp
/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
class Solution {
public:
    int maxPoints(vector<Point> &points) {
        class Utils {
        public:
            long transform(int a, int b) {
                int c = gcd(a, b);
                a /= c;
                b /= c;
                long result = a;
                return (result << 32) | b;
            }

            int gcd(int a, int b) {
                while (b != 0) {
                    int temp = a % b;
                    a = b;
                    b = temp;
                }
                return a;
            }
        };
        if (points.size() < 3) {
            return points.size();
        }
        int result = 0;
        Utils utils;
        for (int i = 0; i < points.size(); i++) {
            Point current = points[i];
            int samePoint = 1;
            int sameVerticalLine = 0;
            unordered_map<long, int> memory;
            for (int j = i + 1; j < points.size(); j++) {
                Point point = points[j];
                if (current.x == point.x && current.y == point.y) {
                    samePoint++;
                } else if (current.x == point.x) {
                    sameVerticalLine++;
                } else {
                    long key = utils.transform(point.y - current.y, point.x - current.x);
                    if (memory.find(key) == memory.end()) {
                        memory[key] = 0;
                    }
                    memory[key]++;
                }
            }
            result = max(result, sameVerticalLine + samePoint);
            for (auto &item : memory) {
                result = max(result, item.second + samePoint);
            }
        }
        return result;
    }
};

File: 149. Max Points on a Line/Python/Solution.py
# Definition for a point.
# class Point(object):
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b


class Solution(object):
    def maxPoints(self, points):
        """
        :type points: List[Point]
        :rtype: int
        """
        def transform(a, b):
            def gcd(a, b):
                while b:
                    a, b = b, a % b
                return a

            c = gcd(a, b)
            a /= c
            b /= c
            return (a << 32) | b

        if len(points) < 3:
            return len(points)
        result = 0
        for i, current in enumerate(points):
            same_point = 1
            same_vertical_line = 0
            memory = {}
            for j in range(i + 1, len(points)):
                point = points[j]
                if current.x == point.x and current.y == point.y:
                    same_point += 1
                elif current.x == point.x:
                    same_vertical_line += 1
                else:
                    key = transform(point.y - current.y, point.x - current.x)
                    if key not in memory:
                        memory[key] = 0
                    memory[key] += 1
            result = max(result, same_vertical_line + same_point)
            if memory:
                result = max(result, max(memory.values()) + same_point)
        return result
